homeassistant/components/elmax/common.py

```diff
@@ -0,0 +1,229 @@
+                "Network error occurred while contacting ELMAX cloud"
+            ) from err
+        except Exception as err:
+            _LOGGER.exception("Unexpected exception")
+            raise HomeAssistantError("An unexpected error occurred") from err
+
+
+class ElmaxEntity(Entity):
+    """Wrapper for Elmax entities."""
+
+    def __init__(
+        self,
+        panel: PanelEntry,
+        elmax_device: DeviceEndpoint,
+        panel_version: str,
+        coordinator: ElmaxCoordinator,
+    ) -> None:
+        """Construct the object."""
+        self._panel = panel
+        self._device = elmax_device
+        self._panel_version = panel_version
+        self._coordinator = coordinator
+        self._transitory_state = None
+
+    @property
+    def transitory_state(self) -> Any | None:
+        """Return the transitory state for this entity."""
+        return self._transitory_state
+
+    @transitory_state.setter
+    def transitory_state(self, value: Any) -> None:
+        """Set the transitory state value."""
+        self._transitory_state = value
+
+    @property
+    def panel_id(self) -> str:
+        """Retrieve the panel id."""
+        return self._panel.hash
+
+    @property
+    def unique_id(self) -> str | None:
+        """Provide a unique id for this entity."""
+        return self._device.endpoint_id
+
+    @property
+    def name(self) -> str | None:
+        """Return the entity name."""
+        return self._device.name
+
+    @property
+    def extra_state_attributes(self) -> Mapping[str, Any] | None:
+        """Return extra attributes."""
+        return {
+            "index": self._device.index,
+            "visible": self._device.visible,
+        }
+
+    @property
+    def device_info(self):
+        """Return device specific attributes."""
+        return {
+            "identifiers": {(DOMAIN, self._panel.hash)},
+            "name": self._panel.get_name_by_user(
+                self._coordinator.http_client.get_authenticated_username()
+            ),
+            "manufacturer": "Elmax",
+            "model": self._panel_version,
+            "sw_version": self._panel_version,
+        }
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._panel.online
+
+    def _http_data_changed(self) -> None:
+        # Whenever new HTTP data is received from the coordinator we extract the stat of this
+        # device and store it locally for later use
+        device_state = self._coordinator.get_endpoint_state(self._device.endpoint_id)
+        if self._device is None or device_state.__dict__ != self._device.__dict__:
+            # If HTTP data has changed, we need to schedule a forced refresh
+            self._device = device_state
+            self.async_schedule_update_ha_state(force_refresh=True)
+
+        # Reset the transitory state as we did receive a fresh state
+        self._transitory_state = None
+
+    @property
```

Comment:
This is the default from the CoordinatorEntity

Reply:
I was not aware of CoordinatorEntity. I took your advice and refactored the code to rely on a CoordinatorEntity rather. That simplified the overall code. Done in #61273